*** muparser_v134/include/muParserDLL.h.ORIG	2011-07-25 19:35:43.578478044 +0200
--- muparser_v134/include/muParserDLL.h	2011-07-25 19:36:07.977416661 +0200
***************
*** 25,39 ****
  #ifndef MU_PARSER_DLL_H
  #define MU_PARSER_DLL_H
  
! #if defined(WIN32) || defined(_WIN32)
!     #ifdef MUPARSERLIB_EXPORTS
!         #define API_EXPORT(TYPE) __declspec(dllexport) TYPE __stdcall
!     #else
!         #define API_EXPORT(TYPE) __declspec(dllimport) TYPE __stdcall
!     #endif
! #else
      #define API_EXPORT(TYPE) TYPE
! #endif
  
  
  #ifdef __cplusplus
--- 25,39 ----
  #ifndef MU_PARSER_DLL_H
  #define MU_PARSER_DLL_H
  
! //#if defined(WIN32) || defined(_WIN32)
! //    #ifdef MUPARSERLIB_EXPORTS
! //        #define API_EXPORT(TYPE) __declspec(dllexport) TYPE __stdcall
! //    #else
! //        #define API_EXPORT(TYPE) __declspec(dllimport) TYPE __stdcall
! //    #endif
! //#else
      #define API_EXPORT(TYPE) TYPE
! //#endif
  
  
  #ifdef __cplusplus
